### ISSUE TITLE
Illegal character error.

### ISSUE THREAD

**Issue by Freyr666** [ocaml/ocaml-lsp#148]
When I use ocaml-lsp with emacs, I have a following issue: when I accidentally type in Russian and then press backspace erasing all the visible characters, I still have invisible bytes left, leading to illegal character error.

Say, when I type "дуе" instead of "let", then press backspace three times, I'm getting three errors: `Illegal character (\131); Illegal character (\181); Illegal character (\180)`. ("дуе" is represented as "\208\180\209\131\208\181", so it seems like it deletes only first byte of each symbol.)

I don't have such an issue with neither other `lsp` implementations, nor `merlin`.

**Comment by Freyr666** [comment]
Also emacs' `count-lines` doesn't see these invisible bytes, so I believe it's in the server's side document only.

**Comment by giltho** [comment]
This looks like we should be using something like Camomile to handle text instead of plain ocaml strings. The encoding is ALWAYS utf8 over LSP

**Comment by Inkbottle007** [comment]
I typed `let _ = дуе`, and I got right away: "Server ocaml-lsp-server:366915 exited with status exit. Do you want to restart it?..." I had the same result after several calls of `lsp-format` with non-ascii comments in https://github.com/ocaml/ocaml-lsp/issues/161.
(By the way, answering yes at that point lead to "LSP :: Sending to process failed with the following error: Process ocaml-lsp-server not running", but that is another story).
So I saved the file, closed `emacs`, restarted `emacs` on that file (only solution I've found so far to restart the server).
And then I typed backspace on the last character of "дуе". Then, beside that the server crashed  again, I got "Unbound constructor д\203е", which, in my humble utf-8 opinion, would mean that it removed the third byte, out of the 6 necessary to form the string.
```
$ echo -n "дуе" | hd -b
00000000  d0 b4 d1 83 d0 b5                                 |......|
0000000 320 264 321 203 320 265                                        
0000006
```
I then saved the file and observe that from the `emacs` point of view, the third character (and not the third byte) has been deleted.
I did do the operation again because it's so much fun, this time I had the message: "\320у", where the "у" is `U+0443 | у | d1 83`, which is consistent with the removal of the second byte of the 2 char / 4 bytes string.
Bottom line: the behavior might be different on my installation, from the one reported above.

**Comment by Khady** [comment]
> The encoding is ALWAYS utf8 over LSP

Actually text is supposed to be utf16 and positions are supposed to be utf8 offsets.

**Comment by giltho** [comment]
Ah you're right @Khady my bad.

I've never had to play with unicode really in OCaml, is Camomille the best lib to do that kind of things ?

**Comment by Freyr666** [comment]
@giltho Could we get away with transcoding the document  to utf16 (and back), say using [uutf recode](https://erratique.ch/software/uutf/doc/Uutf.html#recode). On my machine it transcodes 100Mb string to and from for 4.6 seconds. Then you could count offsets as bytes multiplied by 2.
